File: apps/backend/src/services/prompt.service.ts
import { nanoid } from 'nanoid';
import { eq } from 'drizzle-orm';
import { db } from '../config/database';
import { prompts } from '../db/schema';
import { NotFoundError } from '../utils/errors';

export class PromptService {
  async createPrompt(userId: string, command: string, title: string, content: string) {
    const promptId = nanoid();
    const now = new Date();

    await db.insert(prompts).values({
      id: promptId,
      userId,
      command,
      title,
      content,
      createdAt: now,
      updatedAt: now,
    });

    return this.getPromptById(promptId, userId);
  }

  async getPrompts(userId: string) {
    return db.select().from(prompts).where(eq(prompts.userId, userId));
  }

  async getPromptById(id: string, userId: string) {
    const prompt = await db.select().from(prompts)
      .where(eq(prompts.id, id))
      .limit(1);

    if (prompt.length === 0 || prompt[0].userId !== userId) {
      throw new NotFoundError('Prompt not found');
    }

    return prompt[0];
  }

  async getPromptByCommand(command: string, userId: string) {
    const prompt = await db.select().from(prompts)
      .where(eq(prompts.command, command))
      .limit(1);

    if (prompt.length === 0 || prompt[0].userId !== userId) {
      throw new NotFoundError('Prompt not found');
    }

    return prompt[0];
  }

  async updatePrompt(id: string, userId: string, data: { command?: string; title?: string; content?: string }) {
    await this.getPromptById(id, userId);

    await db.update(prompts)
      .set({ ...data, updatedAt: new Date() })
      .where(eq(prompts.id, id));

    return this.getPromptById(id, userId);
  }

  async deletePrompt(id: string, userId: string) {
    await this.getPromptById(id, userId);
    await db.delete(prompts).where(eq(prompts.id, id));
  }
}

export const promptService = new PromptService();

File: apps/backend-python/src/services/prompt_service.py
import json
from datetime import datetime
from nanoid import generate
from sqlalchemy.orm import Session
from src.db.models import Prompt
from src.utils.errors import NotFoundError


def _serialize(prompt: Prompt) -> dict:
    return {
        "id": prompt.id,
        "userId": prompt.user_id,
        "command": prompt.command,
        "title": prompt.title,
        "content": prompt.content,
        "createdAt": prompt.created_at.isoformat() if prompt.created_at else None,
        "updatedAt": prompt.updated_at.isoformat() if prompt.updated_at else None,
    }


class PromptService:
    def create_prompt(self, db: Session, user_id: str, command: str, title: str, content: str) -> dict:
        now = datetime.utcnow()
        prompt = Prompt(
            id=generate(), user_id=user_id,
            command=command, title=title, content=content,
            created_at=now, updated_at=now,
        )
        db.add(prompt)
        db.commit()
        db.refresh(prompt)
        return _serialize(prompt)

    def get_prompts(self, db: Session, user_id: str) -> list:
        prompts = db.query(Prompt).filter(Prompt.user_id == user_id).all()
        return [_serialize(p) for p in prompts]

    def get_prompt_by_id(self, db: Session, prompt_id: str, user_id: str) -> dict:
        prompt = db.query(Prompt).filter(Prompt.id == prompt_id, Prompt.user_id == user_id).first()
        if not prompt:
            raise NotFoundError("Prompt not found")
        return _serialize(prompt)

    def get_prompt_by_command(self, db: Session, command: str, user_id: str) -> dict:
        prompt = db.query(Prompt).filter(Prompt.command == command, Prompt.user_id == user_id).first()
        if not prompt:
            raise NotFoundError("Prompt not found")
        return _serialize(prompt)

    def update_prompt(self, db: Session, prompt_id: str, user_id: str, data: dict) -> dict:
        prompt = db.query(Prompt).filter(Prompt.id == prompt_id, Prompt.user_id == user_id).first()
        if not prompt:
            raise NotFoundError("Prompt not found")
        for key in ["command", "title", "content"]:
            if key in data:
                setattr(prompt, key, data[key])
        prompt.updated_at = datetime.utcnow()
        db.commit()
        db.refresh(prompt)
        return _serialize(prompt)

    def delete_prompt(self, db: Session, prompt_id: str, user_id: str) -> None:
        prompt = db.query(Prompt).filter(Prompt.id == prompt_id, Prompt.user_id == user_id).first()
        if not prompt:
            raise NotFoundError("Prompt not found")
        db.delete(prompt)
        db.commit()


prompt_service = PromptService()
